prow/bugzilla/client.go

```diff
@@ -205,46 +294,91 @@ func getRootForClone(c Client, bug *Bug) (*Bug, error) {
 		case l == 1:
 			curr = parent[0]
 		case l > 1:
-			curr = parent[0]
-			errs = append(errs, fmt.Errorf("More than one parent found for bug #%d", curr.ID))
+			curr = oldestBug(parent)
 		}
 	}
 	return curr, utilerrors.NewAggregate(errs)
 }
 
 // GetRootForClone returns the original bug.
 func (c *client) GetRootForClone(bug *Bug) (*Bug, error) {
-	return getRootForClone(c, bug)
+	bugCache := NewBugDetailsCache()
+	return getRootForClone(c, bug, bugCache)
 }
 
 // GetAllClones returns all the clones of the bug including itself
 // Differs from GetClones as GetClones only gets the child clones which are one level lower
-func (c *client) GetAllClones(bug *Bug) ([]*Bug, error) {
-	return getAllClones(c, bug)
+func (c *client) GetAllClones(bugID int) (clones []*Bug, bug *Bug, root *Bug, err error) {
+	bugCache := NewBugDetailsCache()
+	return getAllClones(c, bugID, bugCache)
 }
-func getAllClones(c Client, bug *Bug) ([]*Bug, error) {
-	root, err := getRootForClone(c, bug)
+
+func (c *client) GetCachedClones(bugID int) ([]int, bool) {
+	return c.clonesCache.Get(bugID)
+}
+
+func (c *client) SetCachedClones(bugID int, clonesList []int) {
+	c.clonesCache.Set(bugID, clonesList)
+}
+
+func getAllClones(c Client, bugID int, bugCache *BugDetailsCache) ([]*Bug, *Bug, *Bug, error) {
+
+	var bug *Bug
+	g := new(errgroup.Group)
+	g.Go(func() error {
+		var err error
+		bug, err = c.GetBug(bugID)
+		if err != nil {
+			return err
+		}
+		bugCache.Set(bugID, bug)
+		return nil
+	})
+
+	if clonesID, ok := c.GetCachedClones(bugID); ok {
+		for _, cloneID := range clonesID {
+			cloneID := cloneID
+			g.Go(func() error {
+				var clone *Bug
+				clone, ok = bugCache.Get(cloneID)
+				if !ok {
+					var err error
+					clone, err = c.GetBug(cloneID)
+					if err != nil {
+						return err
+					}
+				}
+				bugCache.Set(clone.ID, clone)
+				return nil
+			})
+		}
+	}
+	if err := g.Wait(); err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to get bug details: %v", err)
+	}
+
+	root, err := getRootForClone(c, bug, bugCache)
```

Comment:
Can we parallelize all of this by doing something like:
```
func GetAllClones(id int) error {
 return getAllClones(id, nil, nil
}

func getAllClones(id int, result cache, errGroup *errgroup.Group) error {
       if cache == nil {
               cache = newCache()
       }
       var shouldWait bool
       if errGroup == nil {
               errGroup = newErrGroup()
               shouldWait = true
       }

       var err error
       bug, cacheHasBug := cache.Has(id)
       if !cacheHasBug() {
               bug, err = getBug(id)
               if err != nil {
                       return err
               }
       }
       errGroup.Go(traverseUp(cache.Get(id), cache, errGroup))
       errGroup.Go(traverseDown(cache.Get(id), cache, errGroup))

       if shouldWait {
               return errGroup.Wait()
       }
       return nil
}

func traverseUp(bug Bug, result cache, errGroup *errgroup.Group) error {
       for _, dependsOnID := range bug.DependsOn {
               dependsOnID := dependsOnID
               errgroup.Go(func() error {
                       if result.Has(dependsOnId) {
                               return nil
                       }
                       parent, err := getBug(dependsOn.Id)
                       if err != nil {
                               return err
                       }
                       result.Set(bug)
                       return getAllClones(parent.ID, cache, errGroup)
               })
       }
       return nil
}

func traverseDown(bug Bug, result cache, errGroup *errGroup.group) error {
       for _, childID := range bug.Blocks {
               childID := childID
               errGroup.Go(func() error {
                       if result.Has(child) {
                               return nil
                       }
                       childBug, err := getBug(childID)
                       if err != nil {
                               return err
                       }
                       result.Set(childBug)
                       return getAllClones(childBug.ID)
               })
       }
       return nil
}
```

Reply:
And probably add an incrementing counter to `traverseUp` and store the value of that together with the bug in the cache so we can find out who the root is?

Reply:
Or maybe just use `getOldestBug` to find the root from the final result?

Reply:
Moving into separate PR as discussed offline